blog/2023-03-21-concurrency-in-rest-apis/index.mdx

```diff
@@ -0,0 +1,175 @@
+
+With all the time in the world, we came up with the following design:
+
+* This **will not be** a real-time collaboration where we can see the changes as they happen
+* A user will only be able to update the document if the document is based on the **latest version of the document**
+
+The business was happy with the design proposal, so we started implementing it.
+
+### The user experience
+
+After we finished the implementation, we tested it, by inviting 2 of our colleagues (Jeannie and Frank) to test it out.
+The following is a timeline of Jeanie and Frank's actions and the results:
+
+![Action sequence](./action-sequence.png)
+
+Our boss was happy with the results!
+
+## Design & Implementation
+
+To be able to achieve such user experience, we came up with the following API design:
+
+* We will use **weak ETags** to handle concurrency
+* We will need 2 API operations, one to **save** and one to **get** a document
+
+### Saving a document
+
+We designed saving a document so that it will use an idempotent `PUT` HTTP method. This method will be used to create a new document, as well as to update an existing one.
+Clients are allowed to set the `documentId`.
+
+The following is a sequence diagram of how the API operation was implemented:
+
+```mermaid
+flowchart TD
+    A([Start]) --> B{Request body present?}
+    B -- Yes --> C[Get documentID path param]
+    C --> D[Get document from DB by documentID]
+    D --> E{Document exists?}
+    E -- No --> F[Save new document]
+    F --> G([HTTP 204])
+    E -- Yes --> H[Get If-Match header value from request]
+    H --> I{Value missing?}
+    I -- Yes --> J([HTTP 428 - Precondition required])
+    I -- No --> K{Request ETag == document Etag?}
+    K -- Yes --> L[Update document]
+    L --> G
+    K -- No -->  N([HTTP 412 -Precondition failed])
+    B -- No ----> O([HTTP 400 - Bad request])
+```
+
+:::tip
+Requirements for this sequence are described in the [RFC 7232](https://www.rfc-editor.org/rfc/rfc7232#section-5)
+:::
+
+### Getting a document
+
+Getting a document is far simpler. Using standard `GET` HTTP method, we will return the document with the provided `documentId`. If there's no document with the provided `documentId`,
+we will return `404 Not Found`.
+
+The following is a sequence diagram of how the API operation was implemented:
+
+```mermaid
+flowchart TD
+    A([Start]) --> B[Resolve document by provided documentID]
+    B --> C{Document exists?}
+    C -- No --> D([HTTP 404])
+    C -- Yes --> E([HTTP 200])
+```
+
+## Sample code
+
+On our <u>[GitHub](https://github.com/gran-software-solutions/code-samples/tree/main/concurrency-in-rest-apis)</u> we've published a sample project that implements the described API.
+The sample code is written in <u>[Kotlin](https://kotlinlang.org/)</u> using <u>[Vert.x](https://vertx.io/)</u> toolkit. It requires having <u>[JDK 11](https://adoptopenjdk.net/)</u> or newer installed
+
+Here's how you can run it:
```

Comment:
the "how to install and run" could be inside the repo and in the blog post we could focus on the problem